rospy controller

omichel opened the pull request:
Pure Python ROS node controller (running on Windows, Linux and macOS).
This provides a solution to Windows users for which the ros controller is currently broken.

omichel commented:
test me

cyberbotics-jenkins commented:
Build Passed :ok_hand:
You can safely merge 'feature-rospy-controller' into 'develop'.

omichel commented:
👍

fabienrohrer commented:
Is there no other solution than to fully copy this module in our repository (for example using pip, package, script, etc.)? We usually avoid this at all costs for obvious reasons.

DavidMansolino commented on projects/languages/ros/worlds/ros_python.wbt:
```diff
@@ -0,0 +1,35 @@
+    "learn how to program a robot."
+  ]
+  title "Robot Programming"
+  window "robot_programming"
```
The title, window and info should be updated.

DavidMansolino commented on projects/languages/ros/controllers/ros_python/ros_python.py:
```diff
@@ -0,0 +1,62 @@
+while robot.step(timeStep) != -1 and not rospy.is_shutdown():
+    pub.publish(sensor.getValue())
+    print('Published sensor value: ', sensor.getValue())
+    if message != '':
```
```suggestion
    if message:
```

DavidMansolino commented on docs/guide/using-ros.md:
```diff
@@ -73,13 +78,13 @@ On Linux distribution, you can find it directly at `/etc/hosts`; on macOS, it is
 However, in this case, you need to setup a build configuration to handle both the "catkin\_make" from ROS and the "Makefile" from Webots to have the resulting binary linked both against the Webots "libController" and the "roscpp" library.
 
-This controller can also be implemented in Python by importing both ROS libraries (roslib, rospy) and Webots libraries (controller) in a Webots robot or supervisor controller.
+On Windows, Linux and macOS, such a controller can be implemented in Python by importing both ROS libraries (including [rospy](http://wiki.ros.org/rospy)) and Webots libraries (controller) in a Webots robot or supervisor controller. A simple example of a Python custom ROS controller is provided in the `WEBOTS_HOME/projects/languages/ros/controllers/ros_python/` folder. A `README.md` file provides ...
```
Does not respect the 1 line per sentence rule.

omichel commented:
The Python libraries included in this PR were taken from https://github.com/Wei1234c/ROS_node_on_Windows as now documented in the `README.md`. I am not sure how much they were modified comparing to the original ROS kinetic files. I tried myself to reuse the same files from the ROS melodic distribution but failed to make them work on Windows. Since they are not so big (less than 500 KB once zipped) and we have no plan to synchronize them with a newer version if any, I feel copying them is simpler and safer. I believe the long term evolution of this is moving to ROS2 using the ROS2 RCL library.

👍

omichel commented:
I removed the python dependencies from the repository and added a wget/unzip in the Makefile of the controller to install them. I also made that the ROS libraries are not downloaded and the ROS controller is not build on Windows, as they are broken.
👍

omichel commented:
👍

DavidMansolino commented:
:eyes:

DavidMansolino commented on projects/default/Makefile:
```diff
@@ -44,11 +34,11 @@ endif
 	@-rm -rf controllers/sumo_supervisor/*.pyc
 
-controllers/%.Makefile libraries/%.Makefile: libraries.Makefile controllers/ros/include.Makefile
+controllers/%.Makefile libraries/%.Makefile: $(ROS_DEPENDENCIES)
```
This is not sufficient in case of make clean, as after a make clean (or cleanse) I still have (on Linux):
```
# testing if everything was cleansed...
Supprimerait projects/default/libraries/ros/libapr-1.so.0
Supprimerait projects/default/libraries/ros/libaprutil-1.so.0
Supprimerait projects/default/libraries/ros/libboost_filesystem.so.1.58.0
Supprimerait projects/default/libraries/ros/libboost_regex.so.1.58.0
Supprimerait projects/default/libraries/ros/libboost_system.so.1.58.0
Supprimerait projects/default/libraries/ros/libboost_thread.so.1.58.0
Supprimerait projects/default/libraries/ros/libconsole_bridge.so.0.2
Supprimerait projects/default/libraries/ros/libcpp_common.so
Supprimerait projects/default/libraries/ros/libicudata.so.55
Supprimerait projects/default/libraries/ros/libicui18n.so.55
Supprimerait projects/default/libraries/ros/libicuuc.so.55
Supprimerait projects/default/libraries/ros/liblog4cxx.so.10
Supprimerait projects/default/libraries/ros/librosconsole.so
Supprimerait projects/default/libraries/ros/librosconsole_backend_interface.so
Supprimerait projects/default/libraries/ros/librosconsole_log4cxx.so
Supprimerait projects/default/libraries/ros/libroscpp.so
Supprimerait projects/default/libraries/ros/libroscpp_serialization.so
Supprimerait projects/default/libraries/ros/librostime.so
Supprimerait projects/default/libraries/ros/libxmlrpcpp.so
Supprimerait projects/default/libraries/ros_kinetic_ubuntu16.04.tar.gz
```

DavidMansolino commented:
test me

DavidMansolino commented:
test me

cyberbotics-jenkins commented:
Build Passed :ok_hand:
You can safely merge 'feature-rospy-controller' into 'develop'.

cyberbotics-jenkins commented:
Build Failed :confused:
https://cyberbotics2.cyberbotics.com:8443/job/Webots/335/consoleText

DavidMansolino commented on tests/sources/test_license.py:
```diff
@@ -89,6 +89,8 @@ def setUp(self):
             'projects/default/libraries/vehicle/java',
             'projects/default/libraries/vehicle/python',
+            'projects/languages/ros/controllers/ros_python/kinetic',
+            'projects/languages/ros/controllers/ros_python/python',
```
We should add a gitignore for these folders.

omichel commented:
test me

cyberbotics-jenkins commented:
Build Failed :confused:
https://cyberbotics2.cyberbotics.com:8443/job/Webots/338/consoleText

omichel commented:
test me

cyberbotics-jenkins commented:
Build Passed :ok_hand:
You can safely merge 'feature-rospy-controller' into 'develop'.

omichel commented:
test me

cyberbotics-jenkins commented:
Build Passed :ok_hand:
You can safely merge 'feature-rospy-controller' into 'develop'.

omichel commented:
👍

fabienrohrer commented:
Just tried to compile this branch on Mac, and I get the following errors:

```
# downloading and installing rospy dependencies
# downloading and installing rospy dependencies
checkdir error:  cannot create kinetic
                 Attribute not found
                 unable to process kinetic/.
checkdir error:  cannot create kinetic/dist-packages
                 Attribute not found
                 unable to process kinetic/dist-packages/.
checkdir error:  cannot create kinetic/dist-packages/catkin
                 Attribute not found
                 unable to process kinetic/dist-packages/catkin/.
warning:  cannot set modif./access times for kinetic/dist-packages/catkin/cmake.py
          No such file or directory
warning:  cannot set modif./access times for kinetic/dist-packages/catkin/environment_cache.py
          No such file or directory
warning:  cannot set modif./access times for kinetic/dist-packages/catkin/find_in_workspaces.py
          No such file or directory
warning:  cannot set modif./access times for kinetic/dist-packages/genmsg/base.py
          No such file or directory
warning:  cannot set modif./access times for kinetic/dist-packages/genpy/generator.py
          No such file or directory
checkdir error:  cannot create kinetic/dist-packages/roscpp/msg
                 Attribute not found
                 unable to process kinetic/dist-packages/roscpp/msg/.
warning:  cannot set modif./access times for kinetic/dist-packages/roscpp/srv/_GetLoggers.py
          No such file or directory
warning:  cannot set modif./access times for kinetic/dist-packages/roscpp/__init__.py
          No such file or directory
checkdir error:  cannot create kinetic/dist-packages/rosgraph/impl
                 Attribute not found
                 unable to process kinetic/dist-packages/rosgraph/impl/.
make[3]: *** [kinetic] Error 2
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [python] Error 2
make[2]: *** [debug] Error 2
make[1]: *** [debug] Error 2
make: *** [webots_target] Error 2
```

omichel commented:
Thank you for testing. I believe I fixed the concurrency problem in the Makefile that caused the error you had.

👍

fabienrohrer commented:
Well spotted, it fixed the build on macOS.

DavidMansolino commented:
:eyes:
